File: Common/ChangeReporter.vb
Imports System.Collections.Specialized
Imports System.ComponentModel

Public Class ChangeReporter

    Dim A As New List(Of Object)()

    Public Sub Add(ByVal Obj As Object)
        Dim PropObj As INotifyPropertyChanged,
            CollecObj As INotifyCollectionChanged,
            Enumerable As IEnumerable

        If Obj.GetType().Name = "NotifyingCollection`1" Then
            A.Add(Obj)
        End If

        PropObj = TryCast(Obj, INotifyPropertyChanged)
        If PropObj IsNot Nothing Then
            AddHandler PropObj.PropertyChanged, AddressOf Me.OnPropertyChanged
        End If

        CollecObj = TryCast(Obj, INotifyCollectionChanged)
        If CollecObj IsNot Nothing Then
            AddHandler CollecObj.CollectionChanged, AddressOf Me.OnCollectionChanged
        End If

        Enumerable = TryCast(Obj, IEnumerable)
        If Enumerable IsNot Nothing Then
            For Each O As Object In Enumerable
                Me.Add(O)
            Next
        End If
    End Sub

    Public Sub Remove(ByVal Obj As Object)
        Dim PropObj As INotifyPropertyChanged,
            CollecObj As INotifyCollectionChanged,
            Enumerable As IEnumerable

        PropObj = TryCast(Obj, INotifyPropertyChanged)
        If PropObj IsNot Nothing Then
            RemoveHandler PropObj.PropertyChanged, AddressOf Me.OnPropertyChanged
        End If

        CollecObj = TryCast(Obj, INotifyCollectionChanged)
        If CollecObj IsNot Nothing Then
            RemoveHandler CollecObj.CollectionChanged, AddressOf Me.OnCollectionChanged
        End If

        Enumerable = TryCast(Obj, IEnumerable)
        If Enumerable IsNot Nothing Then
            For Each O As Object In Enumerable
                Me.Remove(O)
            Next
        End If
    End Sub

    Private Sub OnPropertyChanged(sender As Object, e As PropertyChangedEventArgs)
        Me.OnObjectChanged()
    End Sub

    Private Sub OnCollectionChanged(sender As Object, e As NotifyCollectionChangedEventArgs)
        If e.Action <> NotifyCollectionChangedAction.Move Then
            If e.OldItems IsNot Nothing Then
                For Each O As Object In e.OldItems
                    Me.Remove(O)
                Next
            End If
            If e.NewItems IsNot Nothing Then
                For Each O As Object In e.NewItems
                    Me.Add(O)
                Next
            End If
        End If
        Me.OnObjectChanged()
    End Sub

#Region "ObjectChanged Event"
    Public Event ObjectChanged As EventHandler(Of EventArgs)

    Protected Overridable Sub OnObjectChanged()
        RaiseEvent ObjectChanged(Me, EventArgs.Empty)
    End Sub
#End Region

End Class

File: Common/ChangeReporter.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Collections.Specialized;
using System.ComponentModel;

namespace Ks.Common
{
    public class ChangeReporter
    {
        private readonly List<object> A = new List<object>();

        public void Add(object Obj)
        {
            INotifyPropertyChanged PropObj;
            INotifyCollectionChanged CollecObj;
            IEnumerable Enumerable;

            if (Obj.GetType().Name == "NotifyingCollection`1")
            {
                this.A.Add(Obj);
            }

            PropObj = Obj as INotifyPropertyChanged;
            if (PropObj != null)
            {
                PropObj.PropertyChanged += this.OnPropertyChanged;
            }

            CollecObj = Obj as INotifyCollectionChanged;
            if (CollecObj != null)
            {
                CollecObj.CollectionChanged += this.OnCollectionChanged;
            }

            Enumerable = Obj as IEnumerable;
            if (Enumerable != null)
            {
                foreach (var O in Enumerable)
                {
                    this.Add(O);
                }
            }
        }

        public void Remove(object Obj)
        {
            INotifyPropertyChanged PropObj;
            INotifyCollectionChanged CollecObj;
            IEnumerable Enumerable;

            PropObj = Obj as INotifyPropertyChanged;
            if (PropObj != null)
            {
                PropObj.PropertyChanged -= this.OnPropertyChanged;
            }

            CollecObj = Obj as INotifyCollectionChanged;
            if (CollecObj != null)
            {
                CollecObj.CollectionChanged -= this.OnCollectionChanged;
            }

            Enumerable = Obj as IEnumerable;
            if (Enumerable != null)
            {
                foreach (var O in Enumerable)
                {
                    this.Remove(O);
                }
            }
        }

        private void OnPropertyChanged(object sender, PropertyChangedEventArgs e)
        {
            this.OnObjectChanged();
        }

        private void OnCollectionChanged(object sender, NotifyCollectionChangedEventArgs e)
        {
            if (e.Action != NotifyCollectionChangedAction.Move)
            {
                if (e.OldItems != null)
                {
                    foreach (var O in e.OldItems)
                    {
                        this.Remove(O);
                    }
                }
                if (e.NewItems != null)
                {
                    foreach (var O in e.NewItems)
                    {
                        this.Add(O);
                    }
                }
            }
            this.OnObjectChanged();
        }

        public event EventHandler<EventArgs> ObjectChanged;

        protected virtual void OnObjectChanged()
        {
            ObjectChanged?.Invoke(this, EventArgs.Empty);
        }
    }
}
